consensus/src/txn_cache.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_crypto::ed25519::{Ed25519PrivateKey, Ed25519Signature};
+use aptos_crypto::{hash::DefaultHasher, HashValue, PrivateKey, Uniform};
+use aptos_types::account_address::AccountAddress;
+use aptos_types::chain_id::ChainId;
+use aptos_types::transaction::{RawTransaction, Script, SignedTransaction};
+use bcs::to_bytes;
+use concurrent_lru::sharded::LruCache;
+use dashmap::DashMap;
+use dashmap::DashSet;
+use num_traits::Signed;
+use once_cell::sync::Lazy;
+use parking_lot::RwLock;
+use rand::seq::SliceRandom;
+use rand::thread_rng;
+use rayon::prelude::*;
+use rayon::ThreadPool;
+use serde::Serialize;
+use std::convert::{TryFrom, TryInto};
+use std::hash::Hash;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::Instant;
+
+fn generate_txn(id: u64) -> SignedTransaction {
+    let txn: SignedTransaction = SignedTransaction::new(
+        RawTransaction::new_script(
+            AccountAddress::random(),
+            0,
+            Script::new(vec![], vec![], vec![]),
+            0,
+            0,
+            100 * id,
+            ChainId::test(),
+        ),
+        Ed25519PrivateKey::generate_for_testing().public_key(),
+        Ed25519Signature::try_from(&[1u8; 64][..]).unwrap(),
+    );
+    txn
+}
+
+pub struct ConcurrentTxnCache {
+    mysterious_counter: AtomicU64,
+    state: LruCache<[u8; 32], u64>,
+    use_crypto_hash: bool,
+}
+
+impl ConcurrentTxnCache {
+    pub fn new(cache_size: usize, use_crypto_hash: bool) -> ConcurrentTxnCache {
+        ConcurrentTxnCache {
+            mysterious_counter: AtomicU64::new(0),
+            state: LruCache::new(cache_size as u64),
+            use_crypto_hash,
+        }
+    }
+
+    pub fn get_key(&self, item: &SignedTransaction) -> [u8; 32] {
+        if self.use_crypto_hash {
+            let bytes = to_bytes(item).unwrap();
+            let mut hasher = DefaultHasher::new(b"CacheTesting");
+            hasher.update(&bytes);
+            let hash_res: HashValue = hasher.finish();
+            hash_res.get_bytes()
+        } else {
+            let mut ret = [0_u8; 32];
+            let account_bytes = item.sender().into_bytes();
+            let seq_num_bytes = item.sequence_number().to_le_bytes();
+            for i in 0..24 {
```

Comment:
existing experiment use 32-byte cache key, so I'm also taking the first 24 bytes from the address + 8 bytes from seqnum to avoid more changes. but nothing blocks us from taking all the bytes.